auth/src/lib.rs

```diff
@@ -0,0 +1,313 @@
+            source: base_source,
+            ..Default::default()
+        };
+        Ok(Credential {
+            source: Box::new(refreshed_source),
+        })
+    }
+
+    /// Finds a Source from which to create tokens.
+    async fn base_source(
+        config: CredentialConfig,
+    ) -> Result<Box<dyn Source + Send + Sync + 'static>> {
+        // 1: Known environment variable.
+        if let Ok(file_name) = std::env::var(GOOGLE_APPLICATION_CREDENTIALS_ENV) {
+            let source = Credential::file_source(file_name.into(), config).await?;
+            return Ok(source);
+        }
+        // 2: Well-known file.
+        if let Ok(path) = Credential::well_known_file() {
+            if path.exists() {
+                let source = Credential::file_source(path, config).await?;
+                return Ok(source);
+            }
+        }
+        // 3: Check if in an environment with on Google Cloud
+        if metadata::is_running_on_gce().await {
+            let source = ComputeSource::new(ComputeSourceConfig {
+                scopes: config.scopes,
+            });
+            let source = Box::new(source);
+            return Ok(source);
+        }
+
+        Err(Error::new("unable to detect default credentials"))
+    }
+
+    /// Creates a source from a file type credential such as a Service Account
+    /// Key file or a gcloud user credential.
+    async fn file_source(
+        file_path: PathBuf,
+        config: CredentialConfig,
+    ) -> Result<Box<dyn Source + Send + Sync + 'static>> {
+        let contents = tokio::fs::read(file_path).await.map_err(Error::wrap)?;
+        let file: Key = serde_json::from_slice(&contents).map_err(Error::wrap)?;
+        let source: Box<dyn Source + Send + Sync + 'static> = match file.cred_type {
+            "authorized_user" => {
+                let source = UserSource::from_file_contents(
+                    &contents,
+                    UserSourceConfig {
+                        scopes: config.scopes,
+                    },
+                )?;
+                Box::new(source)
+            }
+            "service_account" => {
+                let source = ServiceAccountKeySource::from_file_contents(
+                    &contents,
+                    ServiceAccountKeySourceConfig {
+                        scopes: config.scopes,
+                    },
+                )?;
+                Box::new(source)
+            }
+            _ => {
+                return Err(Error::new(format!(
+                    "unsupported credential type found: {}",
+                    file.cred_type
+                )));
+            }
+        };
+        Ok(source)
+    }
+
+    /// Returns the path to a gcloud user credential.
+    fn well_known_file() -> Result<PathBuf> {
+        let mut path = PathBuf::new();
+        if cfg!(windows) {
+            let appdata = std::env::var(WINDOWS_APPDATA_ENV)
+                .map_err(|_| Error::new("unable to find APPDATA"))?;
```

Comment:
So callers don't have to compare strings, I still think we need an enum in the error type.  We talked about following the pattern in std::io::Error, with ErrorKind:
```
#[non_exhaustive]
pub enum ErrorKind 
```

Reply:
Done.